### PR TITLE
gridding tweaks + propertymaps 

### PR DESCRIPTION
**Pull request by jakeown**
Added region-specific clipping of far left and right channels to better remove frequency-switched dips.  Also added PropertyMaps from GAS, with minor changes related to Keystone directory structure.  Not sure if that belongs here, but thought I would add it since I am using it for Keystone

### CODE REVIEW

File: keystone/baseline.py [hunk 1]
```diff
@@ -110,6 +110,15 @@ def tightWindow(spectrum, spaxis,
         mask[(spaxis < (v0 - outerwindow))] = True
     return(~mask)
 
+def clipper(file_in, clip_lr=[100,100]):
```
Comment (low-sky) [inline keystone/baseline.py]: Left/right doesn't seem like a great description here since velocity vs. frequency can change this.  Can we also split the keywords instead of passing the list?

File: keystone/baseline.py [hunk 2]
```diff
@@ -110,6 +110,15 @@ def tightWindow(spectrum, spaxis,
...
+	# Clip channels off the left and right of cube
+	# clip_lr = [channels to clip from left, channels to clip from right]
+	clip_cube = SpectralCube.read(file_in)
+	x_clip = clip_cube.unmasked_data[clip_lr[0]:-clip_lr[1], :, :]
+	header_clip = clip_cube.header
+	header_clip['NAXIS3']=len(x_clip[:,0,0])
+	sp = SpectralCube(data=x_clip, wcs=clip_cube.wcs, header=header_clip)
+	sp.write(file_in[0:-5] + '_clip.fits', format='fits', overwrite=True)
```
Comment (low-sky) [inline keystone/baseline.py]: I think this can be replaced with `SpectralCube` slicing, namely
```
cube = SpectralCube.read(file_in)
clip_cube = cube[clip_lr[0]:-clip_lr[1], :, :]
clip_cube.write(file_in.replace('.fits','_clip.fits', overwrite=True)
```

File: keystone/gridder.py [hunk 3]
```diff
@@ -206,23 +206,49 @@ def griddata(pixPerBeam=3.5,
     if outdir is None:
         outdir = os.getcwd()
     
+    if region=='NGC2264':
```
Comment (low-sky) [inline keystone/gridder.py]: with this section of the code growing, I would probably prefer to see this whole chunk here pulled out into a script maybe called `postprocess.py` which does clipping based on information in a version controlled text file (maybe in `yaml` format).  This would allow the code to be separate from the reduction parameters contained in the file.  If you think this is a good idea but aren't sure of how exactly to do this, we could work together to lay that out.  Thoughts?

File: keystone/pipeline.py [hunk 4]
```diff
@@ -124,13 +124,13 @@ def reduceSession(session = 1, overwrite=False, release = 'all'):
                 wrapper(region=region, overwrite = overwrite,
                         release=release, obslog = Log,
                         startdate=Log[LogRows][0]['Date'],
-                        enddate=Log[LogRows][0]['Date'])
+                        enddate=Log[LogRows][0]['Date'], **kwargs)
```
Comment (low-sky) [inline keystone/pipeline.py:127]: 👍

**Review by low-sky — CHANGES_REQUESTED** [review]
Overall, these are good basic innovations, but I'm wondering if we should restructure things a bit to make this more supportable.

File: keystone/baseline.py (hunk 1 above)
Comment (jakeown) [reply keystone/baseline.py]: I have changed clip_lr to lowFreq and highFreq to make it more explicit.

File: keystone/baseline.py (hunk 2 above)
Comment (jakeown) [reply keystone/baseline.py]: Ah the power of SpectralCube! I'll use this instead

File: keystone/gridder.py (hunk 3 above)
Comment (jakeown) [reply keystone/gridder.py]: I think pulling this part out would be a good idea.  I'll create a yaml file and the postprocess.py script, along with a way to weave them into the griddata function

**Comment by jakeown** [comment]
Ok all the requested updates have been added and merged to my fork.  I ended up adding postprocess.py, which contains a function that is used in gridder.py to determine the baselineRegion, and a yaml file with the clipping parameters for each region that postprocess.py views.  

Additionally, I added a new baseline fitting method that I recently developed.  It is based on the RAMPS baseline-fitting method: 1) standard deviations are calculated on sliding 31-channel windows centred on each channel 2) the array of standard deviations is used to determine which channels to include in the baseline fit. RAMPS uses all channels with windows that contain less than 3 times the median of the standard deviation array.  The idea being that channels with signal will have a high standard deviation and be removed from the baseline fit.  I found that taking the windows with the lowest standard deviations worked better, so I sort the standard deviation array and use the bottom 20-40 percent to select the baseline-fit channels. 3) Polynomials of order 1-3 are fitted to the baseline-fit channels and the fit with the lowest reduced chi-square value when compared to the original spectrum is subtracted from the spectrum.  

The motivation behind this is that for tiles with cores at significantly different V_LSR values, the ammonia_window in the keystone.baseline.rebaseline function includes some of the emission line channels, resulting in baseline dips:
w48_rebase3

To get around this, I had been using a single order polynomial fit to the channels on the outer edges of the spectra. That works well for most of the regions, but fails for some like this with significant curvature: (same portion of W48 as above) 
w48_rebase1

Here are the results using my my new method for the same region of W48 as above:
w48_rebase_multi

I have tested on the NH3(1,1) cubes for W48, NGC2264, and Rosette, and the baselines look good.  The function is parallelized, but still took ~2-3 hours for those maps. 

Any thoughts on this approach? I like that it doesn't assume any V_LSR value and simultaneously fits 1-3 order polynomials, which are the main advantages over the existing keystone method.

File: keystone/rebaseline_multi.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+   
+ y = input spectra
+ baselineIndex = indices of spectra to include in baseline fitting
+ blorder_max = largest order polynomial to fit (fit from blorder_max down to order of 1) 
+ noiserms = rms noise of spectrum"""
+    x = np.linspace(-1, 1, len(y))
+    if noiserms is None:
+        noiserms = mad1d((y - np.roll(y, -2))[baselineIndex])
+    out = []
+    for i in numpy.arange(1,blorder_max+1):
```
Comment (low-sky) [inline keystone/rebaseline_multi.py:105]: I don't think this loop is necessary. The point in using Legendre polynomials is that each order is orthogonal. If you fit a cubic, you get coefficients c_3, c_2, c_1, c_0. The corresponding second order fit is just c_2, c_1, c_0. Maybe we need to pass out the coefficient map vs. order to make reconstructing lower order fits easier?

File: keystone/rebaseline_multi.py [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+ blorder = order of polynomial to fit
+ ydata = spectrum y values
+ xdata = spectrum x values
+ blindex = the indices of the spectra to include in baseline fitting
+ noise = rms noise of spectrum"""
+    opts = lsq(legendreLoss, np.zeros(blorder + 1), args=(ydata[blindex],
+                                                          xdata[blindex],
+                                                        noise), loss='arctan')
+    ymod = legendre.legval(xdata, opts.x)
+    chi = redchisqg(ydata,ymod,deg=blorder+1)
```
Comment (low-sky) [inline keystone/rebaseline_multi.py:89]: apropos of my comment on `robustBaseline_chi` I guess you would need to loop over order here from max coeffs down and return an array of chisq value.  I'm guessing it would be more efficient to put the looping at this level to keep the `lsq` calls down to one.  But the robust fitting makes this non-obvious??

File: keystone/rebaseline_multi.py [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+ Function calculates the standard deviation of a 31 pixel window
+ centred on each pixel. A percentage of the pixels with the lowest 
+ standard deviation for their window are then chosen for baseline fitting.    
+   
+ spectra = input spectra as numpy array
+ mask_percent = percentage of pixels to select for baseline fitting    
+       	""" 
+	spec_len = len(spectra)
+	sample = int(spec_len*mask_percent)
+	left = numpy.zeros(15)+numpy.std(spectra[0:31]) # For the first 15 entries, use first window
```
Comment (low-sky) [inline keystone/rebaseline_multi.py:27]: `numpy` to `np` maybe.  
Should this be a `nanstd` throughout?

File: keystone/rebaseline_multi.py [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+	right = numpy.zeros(15)+numpy.std(spectra[-31:]) # For the last 15 entries, use last window
+	middle = numpy.std(skimage.util.view_as_windows(spectra, 31, 1),axis=1)
```
Comment (low-sky) [inline keystone/rebaseline_multi.py:29]: This is cute; I never thought of this approach here.

**Review by low-sky — APPROVED** [review]
Looks good.  Some cosmetic comments, but this should run without further changes.

**Comment by low-sky** [comment]
@jakeown -- The baseline routine clearly works well for us here, so I think this is a great idea.  It would be nice if 31 were a keyword as I'm not a huge fan of hardwiring in constant, but this is for the future. 

I think I'd like to move the multi-processing of the baselining upstream into `gbtpipe`.  Would you be okay with that?  I'd handle the changes there unless you want to PR them in.  Similarly, the signal-identification baseline windows should also be incorporated upstream.

I'm going to go ahead and hit merge soon.  Let me know if I should hit the brakes.

File: keystone/rebaseline_multi.py [hunk 9]
```diff
@@ -0,0 +1,173 @@
...
+ Returns a baseline subtracted spectrum, based on the best-fitting polynomial 
+ for a range of polynomial orders.    
+   
+ y = input spectra
+ baselineIndex = indices of spectra to include in baseline fitting
+ blorder_max = largest order polynomial to fit (fit from blorder_max down to order of 1) 
+ noiserms = rms noise of spectrum"""
+    x = np.linspace(-1, 1, len(y))
+    if noiserms is None:
+        noiserms = mad1d((y - np.roll(y, -2))[baselineIndex])
```
Comment (low-sky) [inline keystone/rebaseline_multi.py:103]: Things are clearly working and this isn't incredibly critical, but the noiserms should be divided by sqrt(2) since the difference between two Gaussian deviates has a distribution that is sqrt(2) larger than the contributing distributions.

File: keystone/rebaseline_multi.py (hunk 5 above)
Comment (jakeown) [reply keystone/rebaseline_multi.py:105]: Yes, this is just me forgetting how Legendre polynomials work.  Removing that loop should reduce the run-time considerably, since I think this was eating up time.  I do think we need to pass out the coefficients.  I am curious how often a third-order fit is preferred by chi squared, for instance

File: keystone/rebaseline_multi.py (hunk 7 above)
Comment (jakeown) [reply keystone/rebaseline_multi.py:27]: That would be better. I was assuming all spectra with at least one non-nan value contained all real signals.

**Comment by jakeown** [comment]
Thanks for the merge. And yes, feel free to move what you need into gbtpipe. I can make these quick changes today and start a new PR here though

File: keystone/rebaseline_multi.py (hunk 9 above)
Comment (jakeown) [reply keystone/rebaseline_multi.py:103]: Ok I see now that sqrt(2) term is in the keystone.baseline.rebaseline function.  I used the keystone.baseline.robustBaseline as the template for this function.  There the sqrt(2) term isn't included when the noiserms parameter isn't supplied.  Is that by design? Or should the sqrt(2) term be there also?

Comment (low-sky) [reply keystone/rebaseline_multi.py:103]: Thanks.  I fixed this too.
